pub-post-iati-identifiers-exist/index.js

```diff
@@ -0,0 +1,132 @@
+        params: {rows: 0},
+        facet: {
+            "identifiers" : {
+                    type: "terms",
+                    field : "iati_identifier",
+                    offset: 0,
+                    limit: 1000
+                }
+            }
+        };
+
+    const solrJsonResponse = await query(queryUrl, 'JSON', false, false, 
+                                         "application/json",
+                                         JSON.stringify(requestBody));
+
+    const result = {};
+
+    if ('identifiers' in solrJsonResponse.facets) {
+        solrJsonResponse.facets.identifiers.buckets.forEach(facetItem => {
+            result[facetItem.val] = facetItem.count;
+        });
+    }
+
+    return result;
+}
+
+export default async function pubIatiIdentifiersExist(context, req) {
+    try {
+        const { body } = req;
+        const queryUrl = new URL(`${config.SOLRCONFIG.url}solr/activity/select`);
+        const BATCH_SIZE = config.SOLR_MAX_ROWS;
+        const MAX_IDS_TO_TEST = config.IATI_IDENTIFIERS_EXIST_MAX_NUMBER_OF_IDS;
+
+        // No body, or body not valid JSON
+        if (!body || JSON.stringify(body) === '{}' || typeof(body) === 'string') {
+            context.res = {
+                status: 400,
+                headers: { 'Content-Type': 'application/json' },
+                body: { error: 'No JSON body' },
+            };
+
+            return;
+        }
+
+        // required keys
+        ['iati_identifiers'].forEach((key) => {
+            if (!(key in body)) {
+                context.res = {
+                    status: 400,
+                    headers: { 'Content-Type': 'application/json' },
+                    body: { error: `Body must contain key "${key}"` },
+                };
+            }
+        });
+        if (context.res.status === 400) {
+            return;
+        }        
+
+        // iatiIdentifiers must be an array
+        if (toString.call(body.iati_identifiers) !== '[object Array]') {
+            context.res = {
+                status: 400,
+                headers: { 'Content-Type': 'application/json' },
+                body: { 
+                    error: '"iati_identifiers" must be an Array of strings (each one an IATI Identifier) ' 
+                         + `but it is of type: ${toString.call(body.iati_identifiers)}`,
+                },
+            };
+            return;
+        }
+
+        const responseBody = {
+            "message": "OK",
+            "detail": "",
+            "unique_iati_identifiers": 0,
+            "total_iati_identifier_occurrences": 0,
+            "iati_identifiers_found": {}
+        };
+
+        // only make the request to Solr if API consumer has given a non-empty list
+        if (body.iati_identifiers.length > 0) {
+            let uniqueIdentifiers = [...new Set(body.iati_identifiers)];
+            if (uniqueIdentifiers.length > MAX_IDS_TO_TEST) {
+                responseBody.message = "Warning";
```

Comment:
Hmmm .... if the user does send more than is allowed, is it better to carry on and risk the user not realising only some were checked or just to throw a 400 straight away and force the user to deal with this? 

I'm just concerned that it's easy for someone to not fully read the docs, to test the function with a small number of ID's, not realise they should also check the "message" key and thus write an app that then misses data when to many are sent at once.

Reply:
Yes, good point. Shall I see what @robredpath thinks?

Reply:
My instinct is that we should quickly and obviously reject requests to that API that we don't intend to fully complete. Given that IATI data can be quite voluminous and omissions hard to spot, I would rather break someone's stuff entirely when they hit this limit than return a partial result.